pkg/providers/p11.go

```diff
@@ -504,32 +504,137 @@ func (p *P11) VerifyCertChain(ctx context.Context, request *istio.VerifyCertChai
+	The length of the chain must be at least 2 when we're done
+	*/
+
+	var retrievedRootCert *x509.Certificate
 
 	var verifyOpts = x509.VerifyOptions{
 		Roots: x509.NewCertPool(),
+		Intermediates: x509.NewCertPool(),
 	}
 
 	if nil == p.cid {
 		err = fmt.Errorf("no loaded CA cert for verification")
 		return
 	}
-	// Todo - do we want to record the label/serial during import too?
-	var retrievedRootCert *x509.Certificate
-	if retrievedRootCert, err = p.ctx.FindCertificate(p.cid, nil, nil); nil != err {
+
+	if 0 != len(request.Certificates) {
+		parsedTargetCert, err = x509.ParseCertificate(request.Certificates[len(request.Certificates)-1])
+		if nil != err {
+			return
+		}
+	} else {
+		err = fmt.Errorf("no certificates supplied")
 		return
 	}
 
-	verifyOpts.Roots.AddCert(retrievedRootCert)
+	switch len(request.Certificates) {
+	case 1:
+		// Try to find a workable CA cert in the HSM
+		if retrievedRootCert, err = p.ctx.FindCertificate(p.cid, nil, nil); nil != err {
+			return
+		}
+		verifyOpts.Roots.AddCert(retrievedRootCert)
+	default:
+		{
+
+		    /*
+		      We try to verify the chain as supplied - if this verifies we then look at the returned chain root and see
+		      if matches our existing root cert
+		     */
+			var parsedFirstCert *x509.Certificate
+
+			if parsedFirstCert, err = x509.ParseCertificate(request.Certificates[0]); nil != err {
+				// TODO - RF: unify
+				// try PEM instead
+				var pemFirstCertBlock *pem.Block
+				pemFirstCertBlock, _ = pem.Decode(request.Certificates[0])
+				parsedFirstCert, err = x509.ParseCertificate(pemFirstCertBlock.Bytes)
+				if nil != err {
+					return
+				}
+			}
+
+			var preliminaryVerifyOpts = x509.VerifyOptions{
+				Roots: x509.NewCertPool(),
+				Intermediates: x509.NewCertPool(),
+			}
+			preliminaryVerifyOpts.Roots.AddCert(parsedFirstCert)
+
+			// And add any supplied intermediate certs
+			for i := 1; i < len(request.Certificates) - 1; i++ {
+
+				var parsedAdditionalIntermediateCert *x509.Certificate
+				if parsedAdditionalIntermediateCert, err = x509.ParseCertificate(request.Certificates[i]); nil != err {
+					logrus.Errorf("failed to parse additional intermediate certificate")
+					return
+				}
+				preliminaryVerifyOpts.Intermediates.AddCert(parsedAdditionalIntermediateCert)
+			}
+
+			var parsedChains [][]*x509.Certificate
+			if parsedChains, err = parsedTargetCert.Verify(preliminaryVerifyOpts); nil != err {
+				logrus.Errorf("supplied chain does not verify")
+                return
+			} else {
+
+				/*
+				Here we examine the verified chains, as yet ignoring our CA certs.
+				If the verified chain root matches our CA cert, all is good
+
+				If not, we treat it as an intermediate cert and proceed to a verification which takes this into account
+
+				For now, we should only have a single chain, so crash out if there's more than one
+				 */
+				if 1 != len(parsedChains) {
+					err = fmt.Errorf("unhandled: multiple verification chains")
```

Comment:
@ProsaicSatsuma would there ever be a use case for multiple chains?  

Maybe a quick word in an issue for a future feature if it's of value?